File: python/mirascope/llm/providers/openai/model_id.py
"""OpenAI model ids, and related utilities."""

from typing import Literal, TypeAlias, get_args

from .model_info import OpenAIKnownModels

OpenAIModelId = OpenAIKnownModels | str
"""Valid OpenAI model IDs including API-specific variants."""

OPENAI_KNOWN_MODELS: set[str] = set(get_args(OpenAIKnownModels))

ApiMode: TypeAlias = Literal["responses", "completions"]


def model_name(model_id: OpenAIModelId, api_mode: ApiMode | None) -> str:
    """Extract the openai model name from the ModelId

    Args:
        model_id: Full model ID (e.g. "openai/gpt-4o")
        api_mode: API mode to append as suffix ("responses" or "completions").
          If None, no suffix will be added (just the base model name).

    Returns:
        Provider-specific model ID with API suffix (e.g. "gpt-4o:responses")
    """
    base_name = (
        model_id.split("/")[1].removesuffix(":responses").removesuffix(":completions")
    )
    if api_mode is None:
        return base_name
    return f"{base_name}:{api_mode}"

File: typescript/src/llm/providers/openai/model-id.ts
/**
 * OpenAI model IDs and related utilities.
 */

import type { OpenAIKnownModels } from "@/llm/providers/openai/model-info";

/**
 * The OpenAI model IDs registered with Mirascope.
 */
export type OpenAIModelId = OpenAIKnownModels | (string & {});

/**
 * API mode for OpenAI requests.
 */
export type ApiMode = "responses" | "completions";

/**
 * Extract the OpenAI model name from the ModelId.
 *
 * @param modelId - Full model ID (e.g. "openai/gpt-4o" or "openai/gpt-4o:responses")
 * @param apiMode - API mode to append as suffix ("responses" or "completions").
 *   If null, no suffix will be added (just the base model name).
 * @returns Provider-specific model ID with API suffix (e.g. "gpt-4o:responses")
 */
export function modelName(
  modelId: OpenAIModelId,
  apiMode: ApiMode | null,
): string {
  const modelPart = modelId.replace(/^openai\//, "");
  const baseName = modelPart
    .replace(/:responses$/, "")
    .replace(/:completions$/, "");

  if (apiMode === null) {
    return baseName;
  }
  return `${baseName}:${apiMode}`;
}
